azurerm/resource_arm_azure_firewall_network_rule_collection.go

```diff
@@ -0,0 +1,355 @@
+
+	azureRMLockByName(firewallName, azureFirewallResourceName)
+	defer azureRMUnlockByName(firewallName, azureFirewallResourceName)
+
+	firewall, err := client.Get(ctx, resourceGroup, firewallName)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	ipConfigs := fixArmAzureFirewallIPConfiguration(&firewall)
+	firewall.AzureFirewallPropertiesFormat.IPConfigurations = &ipConfigs
+
+	ruleCollections := *firewall.AzureFirewallPropertiesFormat.NetworkRuleCollections
+	existingCollection, exists := findArmAzureFirewallNetworkRuleCollectionByName(&firewall, name)
+	if exists {
+		log.Printf("[INFO] updating existing AzureRM Azure Firewall network rule collection")
+		existingCollection.AzureFirewallNetworkRuleCollectionPropertiesFormat.Action = &network.AzureFirewallRCAction{
+			Type: action,
+		}
+		existingCollection.AzureFirewallNetworkRuleCollectionPropertiesFormat.Priority = priority
+		rules := expandArmAzureFirewallNetworkRules(d)
+		existingCollection.Rules = &rules
+	} else {
+		log.Printf("[INFO] adding new AzureRM Azure Firewall network rule collection")
+		newFwRuleCol := expandArmAzureFirewallNetworkRuleCollection(d)
+		ruleCollections = append(ruleCollections, newFwRuleCol)
+	}
+	firewall.AzureFirewallPropertiesFormat.NetworkRuleCollections = &ruleCollections
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, firewallName, firewall)
+	if err != nil {
+		return fmt.Errorf("Error creating/updating Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	err = future.WaitForCompletion(ctx, client.Client)
+	if err != nil {
+		return fmt.Errorf("Error waiting for creation/update of Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	read, err := client.Get(ctx, resourceGroup, firewallName)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read Azure Firewall %q (Resource Group %q) ID", firewallName, resourceGroup)
+	}
+
+	var collectionID string
+	for _, collection := range *read.AzureFirewallPropertiesFormat.NetworkRuleCollections {
+		if *collection.Name == name {
+			collectionID = *collection.ID
+		}
+	}
+	d.SetId(collectionID)
+
+	return resourceArmAzureFirewallNetworkRuleCollectionRead(d, meta)
+}
+
+func resourceArmAzureFirewallNetworkRuleCollectionRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).azureFirewallsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	resourceGroup := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+	fwName := d.Get("azure_firewall_name").(string)
+
+	firewall, err := client.Get(ctx, resourceGroup, fwName)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Azure Firewall %q (Resource Group %q): %+v", name, resourceGroup, err)
```

Comment:
if the firewall returns a 404 (there's a helper you can use) we should remove this from the state, which would allow a re-creation here (whilst this seems odd, this is a nested resource so it's a little different) e.g.

```
if err != nil {
  if utils.ResponseWasNotFound(firewall.Response()) {
    d.SetID("")
    return nil
  }
}
```